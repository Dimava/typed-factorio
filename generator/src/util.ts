export function sortByOrder(a: { order: number }, b: { order: number }): number {
  return a.order - b.order
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function assertNever(_value: never): never {
  throw new Error("Value should be never")
}

export function getFirst<T>(elemOrArray: T | T[]): T {
  if (Array.isArray(elemOrArray)) return elemOrArray[0]
  return elemOrArray
}
