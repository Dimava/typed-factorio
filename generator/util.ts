export function byOrder(a: { order: number }, b: { order: number }): number {
  return a.order - b.order
}

export function assertNever(value: never): never {
  throw new Error("Value should be never: " + value)
}

export function getFirst<T>(elemOrArray: T | T[]): T {
  if (Array.isArray(elemOrArray)) return elemOrArray[0]
  return elemOrArray
}
