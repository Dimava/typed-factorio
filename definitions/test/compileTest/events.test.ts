import { tstl } from "../util"

describe("on_event", () => {
  test("set handler", () => {
    tstl`
      script.on_event(defines.events.on_ai_command_completed, (e: OnAiCommandCompletedEvent) => {
        log(e.result)
      })
      script.on_event("someInput", (e: CustomInputEvent) => {
        log(serpent.block(e))
      })
      script.on_event(defines.events.on_area_cloned, undefined)
    `.expectToMatchSnapshot()

    tstl`
      script.on_event(defines.events.on_area_cloned, (e: OnAiCommandCompletedEvent)=> {})
    `.expectToHaveDiagnostics()
  })

  test("filters", () => {
    tstl`
      script.on_event(
        defines.events.on_built_entity,
        (e: OnBuiltEntityEvent) => {
          log(serpent.line(e))
        },
        [
          {
            filter: "type",
            type: "foo",
          },
          {
            filter: "ghost_type",
            type: "bar",
          },
        ]
      )
    `.expectToHaveNoDiagnostics()

    tstl`
      script.on_event(
        defines.events.on_tick,
        0 as any,
        0 as any
      )
    `.expectToHaveDiagnostics()
  })
})
describe("Event filters", () => {
  test("set_event_filter", () => {
    tstl`
      script.set_event_filter(defines.events.on_robot_built_entity, [
        {
          filter: "circuit-network-connectable",
        },
      ])
    `.expectToHaveNoDiagnostics()

    tstl`
      script.set_event_filter(defines.events.on_tick, 0 as any)
    `.expectToHaveDiagnostics()
  })

  test("get_event_filter", () => {
    tstl`
      const filters: LuaEntityDamagedEventFilter[] | undefined = script.get_event_filter(defines.events.on_entity_damaged)
    `.expectToHaveNoDiagnostics()

    tstl`
      script.get_event_filter(defines.events.on_surface_cleared)
    `.expectToHaveDiagnostics()
  })
})

test("custom event", () => {
  tstl`
    const myCustomEvent = script.generate_event_name<{ myCustomData: number }>()
    script.on_event(myCustomEvent, (d) => {
      log(d.myCustomData)
    })
    script.raise_event(myCustomEvent, { myCustomData: 0 })
    script.get_event_handler(myCustomEvent)!({ myCustomData: 0 })
  `.expectToHaveNoDiagnostics()
})
