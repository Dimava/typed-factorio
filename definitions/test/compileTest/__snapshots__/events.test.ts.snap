// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on_event set handler 1`] = `
"script.on_event(
    defines.events.on_ai_command_completed,
    function(e)
        log(e.result)
    end
)
script.on_event(
    \\"someInput\\",
    function(e)
        log(
            serpent.block(e)
        )
    end
)
script.on_event(defines.events.on_area_cloned, nil)
"
`;
